### PR TITLE
[GEOT-6363]Error in SQL conversion when function uses a like expression

### PR DESCRIPTION
**Pull request by imranrajjad**
JIRA:https://osgeo-org.atlassian.net/browse/GEOT-6363

Fixed issue related to using String functions on database fields through CQL filters.

E.g the below CQL filter was failing on PostGIS feature sources 

`strToLowerCase("state_name") LIKE '%texa%'`

### CODE REVIEW

File: modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java [hunk 1]
```diff
@@ -523,8 +523,9 @@ public Object visit(PropertyIsLike filter, Object extraData) {
 
         // JD: hack for date values, we append some additional padding to handle
         // the matching of time/timezone/etc...
-        AttributeDescriptor ad = (AttributeDescriptor) att.evaluate(featureType);
-        if (ad != null && Date.class.isAssignableFrom(ad.getType().getBinding())) {
+        Class attributeType = getExpressionType(att);
+        // AttributeDescriptor ad = (AttributeDescriptor) att.evaluate(featureType);
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java]: No commented out code, please.

File: modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureSourceOnlineTest.java [hunk 2]
```diff
@@ -690,4 +693,20 @@ public void testMixedEncodeIn() throws Exception {
...
+     * @throws Exception
+     */
+    public void testStringFunction() throws Exception {
+
+        FilterFactory ff = dataStore.getFilterFactory();
+        Function function = ff.function("strToLowerCase", ff.property("stringProperty"));
+
+        // should hit the row where stringProperty starts with z (e.g zero)
+        PropertyIsLike likeWithStringFunction = ff.like(function, "z%", "%", "-", "\\", true);
+        assumeTrue(dataStore.getFilterCapabilities().fullySupports(likeWithStringFunction));
```
Comment (aaime) [inline modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureSourceOnlineTest.java]: Please put assume as the first line in the test, makes the precondition clearer.

File: modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisFilterToSQLTest.java [hunk 3]
```diff
@@ -200,4 +202,14 @@ public void testEncodeInArrayWithCast() throws Exception {
         String sql = writer.toString().toLowerCase();
         assertEquals("where 5::text=any(testarray)", sql);
     }
+    
+    @Test
+    public void testIsLike() throws Exception {
```
Comment (aaime) [inline modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisFilterToSQLTest.java]: TestFunctionLike

File: modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java (hunk 1 above)
Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java]: will remove,

**Comment by imranrajjad** [comment]
@aime thank you for your review, I applied your suggestions, let me know if something else is needed.

**Comment by imranrajjad** [comment]
@aaime strangely moving `assumeTrue` on top in online test is failing on `JDBCFeatureSourceOnlineTest` for H2Datastore

File: modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureSourceOnlineTest.java [hunk 4]
```diff
@@ -690,4 +694,19 @@ public void testMixedEncodeIn() throws Exception {
...
         }
     }
+
+    /**
+     * Online tests for String functions along with Like operator
+     *
+     * @throws Exception
+     */
+    public void testStringFunction() throws Exception {
+        assumeTrue(dataStore.getFilterCapabilities().supports(FilterFunction_strToLowerCase.class));
```
Comment (aaime) [inline modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureSourceOnlineTest.java]: Hum... these tests are based on Junit3, while the assume you pulled is from Junit 4 (the jar contains both). Hmm... maybe it should be done in a different way, like this for example:
https://github.com/geotools/geotools/blob/865cc72a90b07841a4c871d1b07fcc8e5defd35b/modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostGISHStoreOnlineTest.java#L55

Comment (imranrajjad) [reply modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureSourceOnlineTest.java]: understood, on it

**Comment by imranrajjad** [comment]
@aime thank you for your review, I applied your suggestions, let me know if something else is needed.

File: modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisFilterToSQLTest.java [hunk 5]
```diff
@@ -200,4 +201,21 @@ public void testEncodeInArrayWithCast() throws Exception {
         String sql = writer.toString().toLowerCase();
         assertEquals("where 5::text=any(testarray)", sql);
     }
+
+    @Test
+    public void TestFunctionLike() throws Exception {
```
Comment (aaime) [inline modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisFilterToSQLTest.java]: A method starting with a capital letter?

Comment (imranrajjad) [reply modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostgisFilterToSQLTest.java]: will fix it, apologies

**Comment by aaime** [comment]
Merging thanks. Can you do backport PRs?

**Comment by imranrajjad** [comment]
> Merging thanks. Can you do backport PRs?

sure

**Pull request merged**